onset_maproom/calc.py

```diff
@@ -2,51 +2,219 @@
-    dims=onset_date.dims,
-    coords=onset_date.coords,
-    attrs=dict(
-        description="Onset Date",
-    ),
-  )
-#Tip to get dates from timedelta  early_start_day
-#  onset_date = onset_date[time_coord] + onset_date  
-  return onset_date
+    zarr_data = xr.open_zarr(zarr_path)
+    return zarr_data
+
+
+# Growing season functions
+
+
+def onset_date(
+    daily_rain,
+    early_start_day,
+    early_start_month,
+    search_days,
+    rainy_day,
+    running_days,
+    running_total,
+    min_rainy_days,
+    dry_days,
+    dry_spell,
+    time_coord="T",
+):
+    """Function reproducing Ingrid onsetDate function
+    http://iridl.ldeo.columbia.edu/dochelp/Documentation/details/index.html?func=onsetDate
+    with the exception that:
+    output is a random deltatime rather than an actual onset
+    earlyStart input is now 2 arguments: day and month (as opposed to 1)
+    output is now the timedelta with each year's earlyStart (as opposed to the date itself)
+    """
+    onset_date = daily_rain[
+        (daily_rain[time_coord].dt.day == early_start_day)
+        & (daily_rain[time_coord].dt.month == early_start_month)
+    ]
+    onset_date = xr.DataArray(
+        data=np.random.randint(0, high=search_days, size=onset_date.shape).astype(
+            "timedelta64[D]"
+        ),
+        dims=onset_date.dims,
+        coords=onset_date.coords,
+        attrs=dict(
+            description="Onset Date",
+        ),
+    )
+    # Tip to get dates from timedelta  early_start_day
+    #  onset_date = onset_date[time_coord] + onset_date
+    return onset_date
+
+
+# Time functions
+
+
+def daily_tobegroupedby_season(
+    daily_data, start_day, start_month, end_day, end_month, time_coord="T"
+):
+    """Returns dataset ready to be grouped by with:
+    the daily data where all days not in season of interest are dropped
+    season_starts: an array where the non-dropped days are indexed by the first day of their season -- to use to groupby
+    seasons_ends: an array with the dates of the end of the seasons
+    Can then apply groupby on daily_data against seasons_starts, and preserving seasons_ends for the record
+    If starting day-month is 29-Feb, uses 1-Mar.
+    If ending day-month is 29-Feb, uses 1-Mar and triggers the option to open the right edge of the Interval.
+    That means that the last day included in the season will be 29-Feb in leap years and 28-Feb otherwise
+    """
+    # Deal with leap year cases
+    if start_day == 29 and start_month == 2:
+        start_day = 1
+        start_month = 3
+    end_day2 = end_day
+    end_month2 = end_month
+    if end_day == 29 and end_month == 2:
+        end_day2 = 1
+        end_month2 = 3
+    # Find seasons edges
+    start_edges = daily_data[time_coord].where(
+        (
+            (daily_data[time_coord].dt.day == start_day)
+            & (daily_data[time_coord].dt.month == start_month)
+        ),
+        drop=True,
+    )
```

Comment:
A stylistic comment that's not essential; take it or leave it.

Here's some shorthand you could use here and elsewhere to further reduce the amount of copy-paste:
```python
start_edges = daily_data[time_coord].where(
    lambda x: x.dt.day == start_day & x.month = start_month,
    drop=True
)
```

Reply:
how does that work? How does it know that x is daily_data[time_coord]?

Reply:
First of all,
```
lambda x: blahblah
```
 is equivalent to writing `myfunc` when `myfunc` is defined as
```
def myfunc(x):
    return blahblah
```
i.e. the value of the lambda expression is a function.

The [documentation](http://xarray.pydata.org/en/stable/generated/xarray.DataArray.where.html) for `DataArray.where` says:

> If [the first argument is] a callable [e.g. a function], it must expect this object as its only parameter.

So if `da` is a `DataArray`, then `da.where(lambda x: x > 1)` calls the lambda function with `x` set to `da`, and expects it to return a `DataArray` of booleans.